Update login/sign-up API

HazemBittar opened the pull request:
Please check login/sign-up and logout RESTful API using jwt-authentication
closes #3 closes #5

HazemBittar committed:
Update register - login api

HazemBittar committed:
Update User.js

HazemBittar committed:
Update middleware.js

abelRoland commented on api/models/User.js:
```diff
@@ -1,21 +1,97 @@
+    name: {
+        type: String,
+        required: true,
+        unique: true,
```
I am not sure if the name field should be unique.... Two "John Doe" is not a problem, the field that is unique generally is email.

Toinne replied:
I agree shouldn't be unique

HazemBittar replied:
Solved!

abelRoland commented on api/models/User.js:
```diff
@@ -1,21 +1,97 @@
+    });
+}
+
+// generate token
+
+userSchema.methods.generateToken = function(cb) {
+    var user = this;
+    var token = jwt.sign(user._id.toHexString(), confiq.SECRET);
+
+    user.token = token;
+    user.save(function(err, user) {
+        if (err) return cb(err);
+        cb(null, user);
+    })
+}
+
+// find by token
+userSchema.statics.findByToken = function(token, cb) {
+    var user = this;
+
+    jwt.verify(token, confiq.SECRET, function(err, decode) {
+        user.findOne({ "_id": decode, "token": token }, function(err, user) {
+            if (err) return cb(err);
+            cb(null, user);
+        })
+    })
+};
+
+//delete token
+
+userSchema.methods.deleteToken = function(token, cb) {
+    var user = this;
+
+    user.update({ $unset: { token: 1 } }, function(err, user) {
+        if (err) return cb(err);
+        cb(null, user);
+    })
+}
+
+
```
I am not sure if all of this belongs to here... Normally the model file is just a model for the dB, all the rest related with the model is done in external files.

Toinne replied:
I would suggest the same, in a big application.
You don't want to mix auth logic and token within the model those are 2 different concepts.

However, this is a small enough application and I don't see the problem here. 
Mongoose provides this easy-to-use way for this sorta use case.

Sayed94h reviewed (commented):
We should think about it:
What should be really unique?
As we see in another social media app: 
Your have `name`, `family name` and **username** which is unique and you can use as your login credential as well. If there are five Bobs and we want to display their names beside their stories. It is not important if they don't have unique name but if the listeners click on their name and visit their profile then they should see a unique **username** maybe combination of their `firstname`  and `lastname` or if it is still not unique we should add some number or his/her birthdate on it. Ex: Bob Becker --> bobbecker1980.

I hope it is clear. 
Nice work @HazemBittar When other reviewers approve the PR I will approve it as well. I asked other backend developers to review your PR.
Have great times

HazemBittar committed:
fix model

Toinne committed:
User: register and login bugfixes

Create database connection in server.js
Don't sync fake user on every database connect
Fix sync to create users
Don't expose user password
Fix typo's in config
Codestyle remove unessary if's and fix identation

Toinne commented on api/db/mongoose.js:
```diff
@@ -1,17 +1,16 @@
 
 const uri = process.env.DB_URI
 mongoose.connect(uri, {useNewUrlParser: true, useUnifiedTopology: true});
-seedUsers();
```
FIX: we don't want to seed users on every connect, each time we restart the server we make DB calls not needed. This is why we use the sync.js script.

Toinne commented on api/routes/login.js:
```diff
@@ -1,69 +1,105 @@
+            }
+            res.status(200).json({
+                success: true,
+                user: { // Control the output we don't want to return the password that would be bad
```
FIX: this was an important change NEVER expose the password to the frontend. Even though it is hashed this poses a security risk. We make it way harder for the hacker to get the hashes if he needs access to the server db.

Toinne commented on api/routes/login.js:
```diff
@@ -1,69 +1,105 @@
+
+        if (user) {
+            return res.status(400).json({
+                error: true,
```
FIX: cleaned error messages a bit. Keep them consistent each error should have a message keep the same structure throughout the application.

Toinne commented on api/routes/login.js:
```diff
@@ -1,69 +1,105 @@
+// login user
+app.post('/login', function(req, res) {
+    let token = req.cookies.auth;
+    User.findByToken(token, (err, user) => {
```
Would say this is a bit unconventional normally you just verify the token and that is enough.
It expires after a set amount of time and you are the only party that can sign and verify it with your secret.

No need to save it on the model.

If you do want to save it for security reason, giving you control to destroy sessions I would suggest a different collection.

But no need to change now this works for now.

Eltayeb-Elgaali committed:
Merge branch 'main' into Register-api

Eltayeb-Elgaali commented:
Many changes have been done by @Toinne .
So i merge this PR.
If someone see that we need to do new changes we can do it through new ticket.
